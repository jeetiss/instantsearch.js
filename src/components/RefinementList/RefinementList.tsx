/** @jsx h */

import { h } from 'preact';
import { Template as WidgetTemplate } from '../../types';
import SharedRefinementList from '../SharedRefinementList/SharedRefinementList';

interface RefinementListCSSClasses {
  root: string;
  noRefinementRoot: string;
  noResults: string;
  list: string;
  item: string;
  selectedItem: string;
  label: string;
  checkbox: string;
  labelText: string;
  showMore: string;
  disabledShowMore: string;
  count: string;
  searchBox: string;
  searchable: {
    searchableRoot: string;
    searchableForm: string;
    searchableInput: string;
    searchableSubmit: string;
    searchableSubmitIcon: string;
    searchableReset: string;
    searchableResetIcon: string;
    searchableLoadingIndicator: string;
    searchableLoadingIcon: string;
  };
}

interface RefinementListTemplates {
  item: WidgetTemplate<{
    value: string;
    label: string;
    isRefined: boolean;
    count: number;
    attribute: string;
    url: string;
    cssClasses: RefinementListCSSClasses;
  }>;
  showMoreText: WidgetTemplate<{ isShowingMore: boolean }>;
  searchableNoResults: WidgetTemplate;
  reset: WidgetTemplate<{ cssClasses: object }>;
  submit: WidgetTemplate<{ cssClasses: object }>;
  loadingIndicator: WidgetTemplate<{ cssClasses: object }>;
}

interface RefinementListItem {
  value: string;
  label: string;
  count: number;
  isRefined: boolean;
}

interface RefinementListProps {
  attribute: string;
  canToggleShowMore: boolean;
  createURL(value: string): string;
  cssClasses: RefinementListCSSClasses;
  hasExhaustiveItems: boolean;
  isFromSearch: boolean;
  isShowingMore: boolean;
  items: RefinementListItem[];
  searchFacetValues?(query: string): void;
  searchIsAlwaysActive: boolean;
  searchPlaceholder: string;
  showMore: boolean;
  templateProps: {
    templates: RefinementListTemplates;
  };
  toggleRefinement(value: string): void;
  toggleShowMore(): void;
}

function RefinementList(props: RefinementListProps) {
  return (
    <SharedRefinementList
      attribute={props.attribute}
      canToggleShowMore={props.canToggleShowMore}
      createURL={props.createURL}
      cssClasses={props.cssClasses}
      isFromSearch={props.isFromSearch}
      isShowingMore={props.isShowingMore}
      items={props.items}
      searchFacetValues={props.searchFacetValues}
      searchPlaceholder={props.searchPlaceholder}
      showMore={props.showMore}
      templateProps={props.templateProps}
      toggleRefinement={props.toggleRefinement}
      toggleShowMore={props.toggleShowMore}
    />
  );
}

export default RefinementList;
