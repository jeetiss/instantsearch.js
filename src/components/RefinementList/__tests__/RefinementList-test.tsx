/** @jsx h */

import { h } from 'preact';
import { render, fireEvent } from '@testing-library/preact';
import RefinementList from '../RefinementList';

const templates = {
  item({ value, count, isRefined }) {
    return `
  <label>
    <input
      data-testid="checkbox"
      type="checkbox"
      value="${value}"
      ${isRefined ? 'checked' : ''}
    />
    <span data-testid="value">${value}</span>
    <span data-testid="count">${count}</span>
  </label>
  `;
  },
  showMoreText() {
    return `<button data-testid="show-more">Show more</button>`;
  },
  searchableNoResults: 'searchableNoResults',
  searchableReset: 'searchableReset',
  searchableSubmit: 'searchableSubmit',
  searchableLoadingIndicator: 'searchableLoadingIndicator',
  reset: 'reset',
  submit: 'submit',
  loadingIndicator: 'loadingIndicator',
};

const defaultProps = {
  attribute: 'attribute',
  canToggleShowMore: false,
  createURL: () => '#',
  hasExhaustiveItems: true,
  isFromSearch: false,
  isShowingMore: false,
  items: [],
  searchFacetValues: undefined,
  searchIsAlwaysActive: false,
  searchPlaceholder: '',
  showMore: false,
  templateProps: {
    templates,
  },
  toggleRefinement: () => {},
  toggleShowMore: () => {},
  cssClasses: {
    root: 'root',
    noRefinementRoot: 'noRefinementRoot',
    noResults: 'noResults',
    list: 'list',
    item: 'item',
    selectedItem: 'selectedItem',
    label: 'label',
    checkbox: 'checkbox',
    labelText: 'labelText',
    showMore: 'showMore',
    disabledShowMore: 'disabledShowMore',
    count: 'count',
    searchBox: 'searchBox',
    searchable: {
      searchableRoot: 'searchableRoot',
      searchableForm: 'searchableForm',
      searchableInput: 'searchableInput',
      searchableSubmit: 'searchableSubmit',
      searchableSubmitIcon: 'searchableSubmitIcon',
      searchableReset: 'searchableReset',
      searchableResetIcon: 'searchableResetIcon',
      searchableLoadingIndicator: 'searchableLoadingIndicator',
      searchableLoadingIcon: 'searchableLoadingIcon',
    },
  },
};

describe('RefinementList', () => {
  describe('cssClasses', () => {
    it('should add the `root` class to the root element', () => {
      const props = {
        ...defaultProps,
        cssClasses: {
          ...defaultProps.cssClasses,
          root: 'customRoot',
        },
      };

      const { container } = render(<RefinementList {...props} />);
      const root = container.firstChild;

      expect(root).toHaveClass('customRoot');
    });

    it('should set item classes to the refinements', () => {
      const props = {
        ...defaultProps,
        cssClasses: {
          ...defaultProps.cssClasses,
          item: 'customItem',
        },
        items: [{ label: 'foo', value: 'foo', isRefined: true, count: 10 }],
      };

      const { container } = render(<RefinementList {...props} />);
      const item = container.querySelector('li');

      expect(item).toHaveClass('customItem');
    });

    it('should set active classes to the active refinements', () => {
      const props = {
        ...defaultProps,
        cssClasses: {
          ...defaultProps.cssClasses,
          selectedItem: 'customSelectedItem',
        },
        items: [
          { label: 'foo', value: 'foo', isRefined: true, count: 10 },
          { label: 'bar', value: 'bar', isRefined: false, count: 10 },
        ],
      };

      const { container } = render(<RefinementList {...props} />);
      const [selectedItem, nonSelectedItems] = container.querySelectorAll('li');

      expect(selectedItem).toHaveClass('customSelectedItem');
      expect(nonSelectedItems).not.toHaveClass('customSelectedItem');
    });
  });

  describe('items', () => {
    it('should have the correct names', () => {
      const props = {
        ...defaultProps,
        items: [
          { label: 'foo', value: 'foo', isRefined: false, count: 10 },
          { label: 'bar', value: 'bar', isRefined: false, count: 10 },
        ],
      };

      const { container } = render(<RefinementList {...props} />);
      const [firstItem, secondItem] = container.querySelectorAll('li');

      expect(firstItem).toHaveTextContent('foo');
      expect(secondItem).toHaveTextContent('bar');
    });

    it('should correctly set if refined or not', () => {
      const props = {
        ...defaultProps,
        items: [
          { label: 'foo', value: 'foo', isRefined: false, count: 10 },
          { label: 'bar', value: 'bar', isRefined: true, count: 10 },
        ],
      };

      const { queryAllByTestId } = render(<RefinementList {...props} />);
      const [firstItem, secondItem] = queryAllByTestId('checkbox');

      expect(firstItem).not.toHaveAttribute('checked');
      expect(secondItem).toHaveAttribute('checked');
    });
  });

  describe('count', () => {
    it('should pass the count to the templateData', () => {
      const props = {
        ...defaultProps,
        items: [
          { label: 'foo', value: 'foo', count: 42, isRefined: false },
          { label: 'bar', value: 'bar', count: 16, isRefined: false },
        ],
      };

      const { queryAllByTestId } = render(<RefinementList {...props} />);
      const [firstItem, secondItem] = queryAllByTestId('count');

      expect(firstItem).toHaveTextContent('42');
      expect(secondItem).toHaveTextContent('16');
    });
  });

  describe('showMore', () => {
    it('adds a showMore link when the feature is enabled', () => {
      const props = {
        ...defaultProps,
        items: [
          { label: 'foo', value: 'foo', isRefined: false, count: 10 },
          { label: 'bar', value: 'bar', isRefined: false, count: 10 },
          { label: 'baz', value: 'baz', isRefined: false, count: 10 },
        ],
        showMore: true,
        isShowingMore: false,
        canToggleShowMore: true,
      };

      const { queryByTestId } = render(<RefinementList {...props} />);
      const showMoreTemplate = queryByTestId('show-more');

      expect(showMoreTemplate).toBeInTheDocument();
    });

    it('does not add a showMore link when the feature is disabled', () => {
      const props = {
        ...defaultProps,
        items: [
          { label: 'foo', value: 'foo', isRefined: false, count: 10 },
          { label: 'bar', value: 'bar', isRefined: false, count: 10 },
          { label: 'baz', value: 'baz', isRefined: false, count: 10 },
        ],
        showMore: false,
        isShowingMore: false,
      };

      const { queryByTestId } = render(<RefinementList {...props} />);
      const showMoreTemplate = queryByTestId('show-more');

      expect(showMoreTemplate).not.toBeInTheDocument();
    });
  });

  describe('rendering', () => {
    it('without facets', () => {
      const props = {
        ...defaultProps,
      };
      const { container } = render(<RefinementList {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('without facets from search', () => {
      const props = {
        ...defaultProps,
        searchFacetValues: () => {},
        isFromSearch: true,
        searchPlaceholder: 'Search',
        templateProps: {
          templates: {
            ...defaultProps.templateProps.templates,
            item: item => item.value,
            searchableNoResults: x => JSON.stringify(x),
          },
        },
      };
      const { container } = render(<RefinementList {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('with facets', () => {
      const props = {
        ...defaultProps,
        items: [
          {
            label: 'Amazon',
            value: 'Amazon',
            count: 1200,
            isRefined: false,
          },
          {
            label: 'Google',
            value: 'Google',
            count: 1000,
            isRefined: true,
          },
        ],
        templateProps: {
          templates: {
            ...defaultProps.templateProps.templates,
            item: item => item.value,
          },
        },
      };
      const { container } = render(<RefinementList {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('with facets and show more', () => {
      const props = {
        ...defaultProps,
        items: [
          {
            label: 'Amazon',
            value: 'Amazon',
            count: 1200,
            isRefined: false,
          },
          {
            label: 'Google',
            value: 'Google',
            count: 1000,
            isRefined: true,
          },
        ],
        showMore: true,
        isShowingMore: false,
        canToggleShowMore: true,
        templateProps: {
          templates: {
            ...defaultProps.templateProps.templates,
            item: item => item.value,
            showMoreText: x => JSON.stringify(x),
          },
        },
      };
      const { container } = render(<RefinementList {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('with facets and disabled show more', () => {
      const props = {
        ...defaultProps,
        items: [
          {
            label: 'Amazon',
            value: 'Amazon',
            count: 1200,
            isRefined: false,
          },
          {
            label: 'Google',
            value: 'Google',
            count: 1000,
            isRefined: true,
          },
        ],
        showMore: true,
        isShowingMore: false,
        canToggleShowMore: false,
        templateProps: {
          templates: {
            ...defaultProps.templateProps.templates,
            item: item => item.value,
          },
        },
      };
      const { container } = render(<RefinementList {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('with facets from search', () => {
      const props = {
        ...defaultProps,
        items: [
          {
            label: 'Amazon',
            value: 'Amazon',
            count: 1200,
            isRefined: false,
          },
          {
            label: 'Google',
            value: 'Google',
            count: 1000,
            isRefined: true,
          },
        ],
        searchFacetValues: () => {},
        isFromSearch: true,
        searchPlaceholder: 'Search',
        templateProps: {
          templates: {
            ...defaultProps.templateProps.templates,
            item: item => item.value,
            searchableNoResults: x => JSON.stringify(x),
          },
        },
      };
      const { container } = render(<RefinementList {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('should not refine on click on already refined items', () => {
      const toggleRefinement = jest.fn();
      const props = {
        ...defaultProps,
        items: [
          { label: 'foo', value: 'foo', isRefined: false, count: 10 },
          { label: 'bar', value: 'bar', isRefined: true, count: 10 },
          { label: 'baz', value: 'baz', isRefined: false, count: 10 },
        ],
        templateProps: {
          templates: {
            ...defaultProps.templateProps.templates,
            item: item => `
              <label>
                <input type="radio" checked="${item.isRefined}" />
                ${item.value}
              </span>
            `,
          },
        },
      };

      const { container } = render(<RefinementList {...props} />);
      const checkedItem = container.querySelector('.item [checked="true"]')!;

      fireEvent.click(checkedItem);

      expect(toggleRefinement).toHaveBeenCalledTimes(0);
    });
  });
});
