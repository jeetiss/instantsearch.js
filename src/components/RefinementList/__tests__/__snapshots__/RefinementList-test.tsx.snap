// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefinementList rendering with facets 1`] = `
<div>
  <div
    class="root"
  >
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <div>
          Amazon
        </div>
      </li>
      <li
        class="item selectedItem"
      >
        <div>
          Google
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`RefinementList rendering with facets and disabled show more 1`] = `
<div>
  <div
    class="root"
  >
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <div>
          Amazon
        </div>
      </li>
      <li
        class="item selectedItem"
      >
        <div>
          Google
        </div>
      </li>
    </ul>
    <button
      class="showMore disabledShowMore"
      disabled=""
    >
      <button
        data-testid="show-more"
      >
        Show more
      </button>
    </button>
  </div>
</div>
`;

exports[`RefinementList rendering with facets and show more 1`] = `
<div>
  <div
    class="root"
  >
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <div>
          Amazon
        </div>
      </li>
      <li
        class="item selectedItem"
      >
        <div>
          Google
        </div>
      </li>
    </ul>
    <button
      class="showMore"
    >
      {"isShowingMore":false}
    </button>
  </div>
</div>
`;

exports[`RefinementList rendering with facets from search 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="searchBox"
    >
      <div>
        <form
          action=""
          novalidate=""
          role="search"
        >
          <input
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            maxlength="512"
            placeholder="Search"
            type="search"
          />
          <button
            title="Submit the search query."
            type="submit"
          >
            submit
          </button>
          <button
            hidden=""
            title="Clear the search query."
            type="reset"
          >
            reset
          </button>
          <span
            hidden=""
          >
            loadingIndicator
          </span>
        </form>
      </div>
    </div>
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <div>
          Amazon
        </div>
      </li>
      <li
        class="item selectedItem"
      >
        <div>
          Google
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`RefinementList rendering without facets 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  />
</div>
`;

exports[`RefinementList rendering without facets from search 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <div
      class="searchBox"
    >
      <div>
        <form
          action=""
          novalidate=""
          role="search"
        >
          <input
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            maxlength="512"
            placeholder="Search"
            type="search"
          />
          <button
            title="Submit the search query."
            type="submit"
          >
            submit
          </button>
          <button
            hidden=""
            title="Clear the search query."
            type="reset"
          >
            reset
          </button>
          <span
            hidden=""
          >
            loadingIndicator
          </span>
        </form>
      </div>
    </div>
    <div
      class="noResults"
    >
      {}
    </div>
  </div>
</div>
`;
